data_structures/indexed_priority_queue.cpp

```diff
@@ -0,0 +1,278 @@
+/**
+ * @file
+ * @brief This is an implementation of an min indexed priority queue,
+ * which, unlike a regular priority queue, has key value pairs, which 
+ * can be looked up in O(1), just like a hashmap. lower priority numbers
+ * refer to higher priority. 
+ * @details
+ * As mentioned in the brief, an indexed priority queue operates on 
+ * the basis of a hashmap, providing its benefit of having O(1) searching
+ * and log(N) modification/removal of ANY key in the queue (not only the front)
+ * but also the benefits of a priority queue for polling min values in 
+ * logarithmic time or having its heap properties. The keys in this 
+ * implementation are the unique identifiers or stored values of the 
+ * queue, whilst the values refer to the comparable that will be used
+ * to prioritise different items. 
+ * More information can be found [here](https://en.wikipedia.org/wiki/Priority_queue)
+ * refer to william fiset's video for a video explanation:
+ * https://youtu.be/jND_WJ8r7FE
+ * @author [ron0studios](https://github.com/ron0studios)
+ */
+
+
+#include <vector>          /// for std::vector
+#include <utility>        /// for std::pair
+#include <unordered_map> /// for std::unordered_map
+#include <cmath>        /// for std::floor
+#include <string>       /// for std::string
+#include <cassert>      /// for assert
+#include <iostream>     /// for IO operations
+
+
+/**
+ * @brief This is the class used to instantiate the indexed priority queue
+ * @tparam T1 this is the identifier value for the queue, and is what will be
+ * stored in the data-structure. Please make sure this is a HASHABLE type!
+ * @tparam T2 this is the priority value for the queue, used to prioritise
+ * certain items over other. Please make sure this is both HASHABLE and 
+ * COMPARABLE!
+ */
+template <class T1 = int, class T2 = int>
+class IPQ
+{
+	private:
+		std::unordered_map<T1,T2> val; ///< key : val
+		std::unordered_map<T1,int> pm; ///< key : heap index
+		std::unordered_map<int,T1> im; ///< heap index : key
+		int size = 0; ///< the size of the heap
+		
+	public:
+						
+		/**
+		 * @brief Constructor for IPQ
+		 * @param keyval this is a vector of key value pairs used initially (optional)
+		 */
+        explicit IPQ(std::vector<std::pair<T1, T2>> keyval =
+                         std::vector<std::pair<T1, T2>>()) {
+            for(auto const &i : keyval){
+				insert(i.first, i.second);
+			}
+        }
+
+        /**
+		 * @brief check if the queue contains a certain key
+		 * @param key the key to check for
+	     */

```

Comment:
The commenting style and spacing looks weird to me. Could you please correctly align stuff and follow the style guide?

Reply:
Hi [tjgurwara99](https://github.com/tjgurwara99)! I just made the relevant changes. It looked perfectly fine in my editor, but showed up odd on github, must be a tabs/spacing issue on my end :)